examples/plot_metadata_routing.py

```diff
@@ -0,0 +1,612 @@
+
+
+# %%
+# The above estimator now has all it needs to consume metadata. This is done by
+# some magic done in :class:`~base.BaseEstimator`. There are now three methods
+# exposed by the above class: ``fit_requests``, ``predict_requests``, and
+# ``get_metadata_routing``. There is also a ``score_requests`` for
+# ``sample_weight`` which is provided by :class:`~base.ClassifierMixin`.
+#
+# By default, no metadata is requested, which we can see as:
+
+print_routing(ExampleClassifier())
+
+# %%
+# The above output means that ``sample_weight`` and ``groups`` are not
+# requested, but if a router is given those metadata, it should raise an error,
+# since the user has not explicitly set whether they are required or not. The
+# same is true for ``sample_weight`` in ``score`` method, which is inherited
+# from :class:`~base.ClassifierMixin`. In order to explicitly set request
+# values for those metadata, we can use these methods:
+
+est = (
+    ExampleClassifier()
+    .fit_requests(sample_weight=False)
+    .predict_requests(groups=True)
+    .score_requests(sample_weight=False)
+)
+print_routing(est)
+
+# %%
+# As you can see, now the metadata have explicit request values, one is
+# requested and one is not. Instead of ``True`` and ``False``, we could also
+# use the :class:`~sklearn.utils.metadata_routing.RequestType` values.
+
+est = (
+    ExampleClassifier()
+    .fit_requests(sample_weight=RequestType.UNREQUESTED)
+    .predict_requests(groups=RequestType.REQUESTED)
+    .score_requests(sample_weight=RequestType.UNREQUESTED)
+)
+print_routing(est)
+
+# %%
+# Please note that as long as the above estimator is not used in another
+# meta-estimator, the user does not need to set any requests for the metadata
+# and the set values are ignored, since a consumer does not validate or route
+# given metadata. A simple usage of the above estimator would work as expected.
+
+est = ExampleClassifier()
+est.fit(X, y, sample_weight=my_weights)
+est.predict(X[:3, :], groups=my_groups)
+
+# %%
+# Now let's have a meta-estimator, which doesn't do much other than routing the
+# metadata.
+
+
+class MetaClassifier(MetaEstimatorMixin, ClassifierMixin, BaseEstimator):
+    def __init__(self, estimator):
+        self.estimator = estimator
+
+    def get_metadata_routing(self):
+        # This method defines the routing for this meta-estimator.
+        # In order to do so, a `MetadataRouter` instance is created, and the
+        # right routing is added to it. More explanation follows.
+        router = MetadataRouter(owner=self.__class__.__name__).add(
+            estimator=self.estimator, method_mapping="one-to-one"
+        )
+        return router
+
+    def fit(self, X, y, **fit_params):
+        # meta-estimators are responsible for validating the given metadata
+        # `get_router_for_object` is a safe way to construct a
+        # `MetadataRouter` from the given object.
+        request_router = get_router_for_object(self)
```

Comment:
`get_router_for_object` can return according to docstring a MetadataRouter or a MetadataRequest. I understand that it is the reason you call the output request_router bugI find this naming of the function confusing then.

Reply:
then maybe call it get_routing_for_object to be consistent with what you said above about get_metadata_routing ? routing means MetadataRouter or MetadataRequest ?